### PR TITLE
[Fix #9734] Add new `Style/TopLevelMethodDefinitions` cop

### LINKED ISSUE #9734: add cop for "global method"
## feature request

We have a project were many people contribute that don't know ruby well, they often write methods that are global, I'd like to stop them
this also often happens in things like Rakefile or *.rake files

## Describe the solution you'd like

```ruby
class Foo
  def not_global <- Good
  end
end
def this_is_global <- Bad
end
```

## Describe alternatives you've considered

`git grep '^def ` works, but would be nice if it's in rubocop

### PR DESCRIPTION
**Pull request by tejasbubane**
Closes #9734

-----------------

Before submitting the PR make sure the following are checked:

* [x] The PR relates to *only* one subject with a clear title and description in grammatically correct, complete sentences.
* [x] Wrote [good commit messages][1].
* [x] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Ran `bundle exec rake default`. It executes all tests and runs RuboCop on its own code.
* [x] Added an entry (file) to the [changelog folder](https://github.com/rubocop/rubocop/blob/master/changelog/) named `{change_type}_{change_description}.md` if the new code introduces user-observable changes. See [changelog entry format](https://github.com/rubocop/rubocop/blob/master/CONTRIBUTING.md#changelog-entry-format) for details.

[1]: https://chris.beams.io/posts/git-commit/

### CODE REVIEW

File: lib/rubocop/cop/style/global_method.rb [hunk 1]
```diff
@@ -0,0 +1,46 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module Style
+      # This cop looks for definitions of global methods.
```
Comment (koic) [inline lib/rubocop/cop/style/global_method.rb:6]: Can you document the reason why it is disabled by default?
https://github.com/rubocop/rubocop/issues/9734#issuecomment-826048992

File: config/default.yml [hunk 2]
```diff
@@ -3421,6 +3421,11 @@ Style/FrozenStringLiteralComment:
     - never
   SafeAutoCorrect: false
 
+Style/GlobalMethod:
```
Comment (bbatsov) [inline config/default.yml:3424]: I'm not sure whether "global" is the right terminology for this. I prefer to think of those as "top-level" method definitions. @rubocop/rubocop-core Ideas how to name this are welcome.

File: lib/rubocop/cop/style/global_method.rb [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+      #     end
+      #   end
+      #
+      #   # good
+      #   Struct.new do
+      #     def some_method
+      #     end
+      #   end
+      class GlobalMethod < Base
+        MSG = 'Do not use global methods.'
```
Comment (bbatsov) [inline lib/rubocop/cop/style/global_method.rb:35]: Maybe "Do not define methods at the top-level." instead?

File: lib/rubocop/cop/style/global_method.rb (hunk 1 above)
Comment (bbatsov) [reply lib/rubocop/cop/style/global_method.rb:6]: I'd probably add a bit more rationale about the cop in general here. I guess some people might not realize where those top-level definitions end up and why there are problematic.

**Comment by bbatsov** [comment]
I think you can also use `define_method` at the top-level, which such a cop should probably detect.

File: config/default.yml (hunk 2 above)
Comment (koic) [reply config/default.yml:3424]: > I prefer to think of those as "top-level" method definitions. 

I agree with this. `TopLevelMethodDefinitions` cop may be clear for the term "top-level". There may be other better names.

**Comment by tejasbubane** [comment]
@bbatsov @koic Changed as per your comments.

**Comment by dvandersluis** [comment]
I probably wouldn't want these to be allowed anyways, but I could see an argument for allowing a `defs` node where the receiver isn't `self`, since in that case, the method will end up scoped to an object and not be global.

ie.

```ruby
def MyClass.foo
  # ...
end
```

I'm fine with this being disallowed but anyone have an opinion there?

**Comment by koic** [comment]
I think it can be allowed if a receiver is explicitly specified. I think top-level methods that are implicitly declared to `Object` without specifying a receiver should be prevented.

File: config/default.yml [hunk 4]
```diff
@@ -4632,6 +4632,11 @@ Style/TernaryParentheses:
     - require_parentheses_when_complex
   AllowSafeAssignment: true
 
+Style/TopLevelMethodDefinitions:
```
Comment (bbatsov) [inline config/default.yml:4635]: Let's use singular here (Definition) to be aligned with how most of the cops are named.

**Comment by bbatsov** [comment]
You forgot to update the commit message itself. 

Btw, I'm thinking that it would be hard to add some check whether the file being analyzed is a script (if it has a shebang) and disable to the cop just for scripts, but probably that's an overkill at this point.

**Comment by bbatsov** [comment]
I guess we should also document this in the Ruby style guide. I'm surprised we haven't done so already.

**Comment by tejasbubane** [comment]
@bbatsov Changed to singular and created PR for styleguide.

**Comment by bbatsov** [comment]
Thanks!

**Pull request merged**